### PR TITLE
Move FFI::DataConverter to Ruby

### PR DESCRIPTION
**Pull request by eregon**
* Include DataConverter in StructByReference in Ruby code to simplify ordering.

See #660 for context.

### CODE REVIEW

File: lib/ffi/data_converter.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
+module FFI
```
Comment (eregon) [inline lib/ffi/data_converter.rb]: Should I add a license header? If so, what copyright line should I use?

**Comment by eregon** [comment]
This is now ready for review.

**Comment by larskanis** [comment]
LGTM

**Comment by eregon** [comment]
Can we merge this?

@tduehr Could you review?

I think this doesn't touch performance critical-code, but I'm happy to run some benchmarks if you tell me which ones would be relevant for this.

The behavior seems similar to [the DataConverter class in JRuby](https://github.com/jruby/jruby/blob/4e8bb2666a7257f0f5986800f96bb88efdd6acbd/core/src/main/java/org/jruby/ext/ffi/DataConverter.java), except that one has the extra `reference_required` and `reference_required?` methods.

**Comment by tduehr** [comment]
There's more to it than just the module itself. I'll take a look at usage today.

I expect this code is used in converting from ruby to C/JAva and back... Any benchmark covering that should be good.

Can you get the benchmark suite working, or at least find the holes? That'd be awesome too...

File: lib/ffi/data_converter.rb (hunk 1 above)
Comment (headius) [reply lib/ffi/data_converter.rb]: Follow whatever pattern other .rb files use. We can discuss sorting out licensing, headers, and copyrights in a separate issue if necessary.

**Comment by headius** [comment]
@tduehr Even if this is on the hot path it's probably a lateral move changing this code to Ruby. It may even be faster in some cases since the C code (and the Java code, for that matter) does uncached instance variable lookups.

**Comment by tduehr** [comment]
The licensing is all over the place. We'll want to consolidate that at some point.

**Comment by eregon** [comment]
I added the license header.

File: lib/ffi/data_converter.rb (hunk 1 above)
Comment (eregon) [reply lib/ffi/data_converter.rb]: I used the same as in `lib/ffi/errno.rb`.

**Pull request merged**